icechunk/src/store.rs

```diff
@@ -337,6 +339,57 @@ impl Store {
         }
     }
 
+    pub async fn delete_dir(&self, prefix: &str) -> StoreResult<()> {
+        if self.read_only().await {
+            return Err(StoreError::ReadOnly);
+        }
+
+        // TODO: Handling preceding "/" is ugly!
+        let path = format!("/{}", prefix.trim_start_matches("/"))
+            .try_into()
+            .map_err(|_| StoreError::BadKeyPrefix(prefix.to_owned()))?;
+        let res = {
+            let guard = self.session.read().await;
+            let node = guard.get_node(&path).await;
+            match node {
+                Ok(NodeSnapshot {
+                    node_data: NodeData::Group, path: node_path, ..
+                }) => Ok((NodeType::Group, node_path)),
+                Ok(NodeSnapshot {
+                    node_data: NodeData::Array(..),
+                    path: node_path,
+                    ..
+                }) => Ok((NodeType::Array, node_path)),
+                Err(err) => Err(err),
+            }
+        };
+
```

Comment:
I think it would be safer to lock the session for writes at the top. Otherwise you need to think about the edge case of somebody else deleting the node (or creating it) between the previous `get_node` and the following `delete_*`.

Reply:
Yes that is the approach I took at first but `store.delete` will try to grab a write lock too. Now I realize I can solve that with a `self.delete_with_optional_locking`.